Extend sys/bitarray API with set-bit counting, xor, nth-bit-set

luqasn opened the pull request:
Hello there!

I am currently working on integrating a memory-optimized forward error correction algorithm for LoRaWAN "natively" in Zephyr instead of relying on the external module that is currently being used.

For that, I need some more bitarray methods than are currently available.
In order to make review more enjoyable, I have added the three additional methods I need in this first PR and will then follow up with a second PR adding the new forward error correction algorithm that uses them.

This is my first Zephyr contribution and I am very much looking forward to your feedback.
Thanks for having a look!

luqasn committed:
lib: bitarray: add method to count bits set in region

This is part one of several changes to add more methods to the bitarray api
so that it can be used for broader usecases, specifically LoRaWAN forward
error correction.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

luqasn committed:
lib: bitarray: add method to xor two bitarrays

This is part one of several changes to add more methods to the bitarray api
so that it can be used for broader usecases, specifically LoRaWAN forward
error correction.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

luqasn committed:
lib: bitarray: add method to find nth bit set in region

This is part one of several changes to add more methods to the bitarray api
so that it can be used for broader usecases, specifically LoRaWAN forward
error correction.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

cfriedt reviewed (changes requested):
Looks like there are still some failing tests

andyross reviewed (dismissed):
Seems clean, though sort of weirdly non-orthogonal.  The original use case for this code was a synchronized slab-ish allocator.  Not sure what xor or popcount are for.  But they fit and don't make anything worse, so no reason not to merge.

luqasn commented:
Thanks for the reviews and sorry for the oversight in the tests.

I added the fix as a new commit as to not have you re-review everything, but please let me know if you want me to fix up the original commit instead.

luqasn commented:
I fixed the test that was failing because of a real implementation gap, but now I got a target dependent failure that I don't understand.

Here is my interpretation:

The tests seem to fail only on mobile cpus due to a check when unlocking a spinlock.
It tries to check if the unlocking cpu is the same as the cpu holding the lock.
But while trying to do that it fails on an assert because apparently getting the current cpu is not supported if `z_smp_cpu_mobile`?
What baffles me is that all other tests pass happily even though they also use spinlocks in the same way.
The only difference I can see is that the method under test for the failure uses two spinlocks instead of one, but that does not seem to be the root of the problem.

Can you help shed some light on this?

andyross commented:
> But while trying to do that it fails on an assert because apparently getting the current cpu is not supported if z_smp_cpu_mobile?

Something got borked.  The assert there is to detect common race conditions with the use of the "current CPU" in contexts where the current thread might migrate to another CPU.  Basically it wants to see that you're in an ISR or that interrupts are masked locally.

If you're holding a spinlock (because you're trying to unlock one) then interrupts really should be masked.  That they aren't  implies that something went wrong.  Most likely you have a mismatched lock somehow, but if you're sure the locking is correct then something very weird is going on.  What platform is this failing on?

luqasn commented:
Thanks for your help!
It failing on some platforms but not on others. One that I looked at the GHA run for was Cortex A53.

I think I will start by isolating the problematic commit/test into a new PR to rule out that one of the test that might run before it are the problem.

And then I'll have to stare long and hard at the code to see if my locking is weird somewhere.

Already did that some and my first idea would be to change the order of the double unlock to unlock in reverse order of the locking to see if that helps. Feels more "correct" anyways.

TaiJuWu commented on lib/utils/bitarray.c:
```diff
@@ -210,6 +210,112 @@ static void set_region(sys_bitarray_t *bitarray, size_t offset,
+	struct bundle_data bd;
+	int ret;
+
+	key = k_spin_lock(&bitarray->lock);
+
+	__ASSERT_NO_MSG(bitarray != NULL);
+	__ASSERT_NO_MSG(bitarray->num_bits > 0);
```
There is a logical issue in my mind.
You should check `bitarray` is not empty before `bitarray->lock`.

I know other function is written like you, but I infer it is a mistake.
Maybe we need others to confirm.

There is another reason I think it is a mistake.
Consider follow scenario:
If a thread take a lock and then assert, it will hold lock forever

luqasn replied:
Yes, I just copy/pasted the structure from the existing methods.

I see what you mean and agree that it is wrong in theory, but I am not sure how important it is.
I think the process will quit anyways, either with a SEGFAULT or with an assertion error.
And if the process exits, it doesn't matter that the lock isn't released.

But if the others agree, I can change all locking in this file to be more theoretically correct.

luqasn commented:
Glad to report that changing the unlock order to be in reverse lock order indeed fixed the issue. 🥳

andyross commented on lib/utils/bitarray.c:
```diff
@@ -311,8 +311,8 @@ int sys_bitarray_xor(sys_bitarray_t *dst, sys_bitarray_t *other, size_t num_bits
 out:
-	k_spin_unlock(&dst->lock, key_dst);
 	k_spin_unlock(&other->lock, key_other);
+	k_spin_unlock(&dst->lock, key_dst);
```
Bingo.  FWIW best to squash this into the original patch.  No need to have the development history of code outside the tree reflected in git.

cfriedt commented:
@luqasn - CI is passing \o/

The only remaining thing to do is convert your last two commits to fixups and squash as Andy mentioned.

something like this should do it (please check my commit math though)
```shell
git reset HEAD~2
# add just the find-nth changes
git commit --fixup d85be1c
# add just the xor changes
git commit --fixup 52b7daa
git rebase -i --autosquash HEAD~5
```

luqasn commented:
Thanks for all the feedback and help, I've squashed and rebased the fixes into their respective commits.

Ready to merge from my side, but what do the others think about @TaiJuWu's remarks?
Should I take care of the erroneous "struct access before NULL check" in the whole file or not?
In this PR or in a new one?

andyross reviewed (approved):
I'm out of complaints, thanks

cfriedt reviewed (changes requested):
Sorry - kind of foolish of me to miss this, but we will definitely need tests for each new function as well.

Ideally they would exercise a meaningful permutation of input parameters for each function, even invalid inputs like `NULL`, etc, but generally we want to exercise as many code paths as possible for coverage.

Usually the order of commits is important.

1) feature A
2) test A
3) feature B
4) test B

martinjaeger commented:
@cfriedt There are tests for each new function, just not added as extra commits, but as part of the commit with the added function itself. I think that's fine.

cfriedt reviewed (approved):
Ah sorry - now I see the tests. Normally we prefer tests in separate commits, but this should be ok

kartben commented on lib/utils/bitarray.c:
```diff
@@ -210,6 +210,112 @@ static void set_region(sys_bitarray_t *bitarray, size_t offset,
+		 * count all bits.
+		 */
+		*count = 0;
+		*count += POPCOUNT(bitarray->bundles[bd.sidx] & bd.smask);
+		*count += POPCOUNT(bitarray->bundles[bd.eidx] & bd.emask);
```
just curious, would this work on all supported toolchains? It looks like the POPCOUNT macro is only defined in toolchain/gcc.h, and not in toolchain/llvm.h, etc.?
